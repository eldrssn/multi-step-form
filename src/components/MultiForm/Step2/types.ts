export type TPlan = Record<string, string>;
