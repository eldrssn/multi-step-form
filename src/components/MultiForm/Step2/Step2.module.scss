@import '/src/styles/colors.scss';
@import '/src/styles/breakpoints.scss';

.buttons {
	position: absolute;
	bottom: 0;

	width: 100%;
	display: flex;
	justify-content: space-between;
}

.form {
	display: flex;
	flex-direction: column;
	gap: 24px;
}

.plans {
  display: flex;
  justify-content: space-between;
  gap: 17px;
  margin-bottom: 30px;
}

.period {
  display: flex;
  background-color: $alabaster;
  justify-content: center;
  align-items: center;
  gap: 25px;
  padding: 15px 0;
}

.option {
  font-size: 14px;
  opacity: 50%;

  &.choosen {
    opacity: 100%;
  }
}

@media screen and (max-width: $mobile) {
  .plans {
    flex-direction: column;
    gap: 10px;
  }
  .buttons {
    position: fixed;
    bottom: 0;
    left: 0;
    right: 0;
    background-color: $white;
    padding: 15px;
  }
}