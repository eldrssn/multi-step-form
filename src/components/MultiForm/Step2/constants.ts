import { TPlan } from './types';

export const plans: TPlan[] = [
  {
    icon: '/src/assets/images/icon-arcade.svg',
    title: 'Arcade',
    priceMonthly: '9',
    priceYearly: '90',
  },
  {
    icon: '/src/assets/images/icon-advanced.svg',
    title: 'Advanced',
    priceMonthly: '12',
    priceYearly: '120',
  },
  {
    icon: '/src/assets/images/icon-pro.svg',
    title: 'Pro',
    priceMonthly: '15',
    priceYearly: '150',
  },
];
