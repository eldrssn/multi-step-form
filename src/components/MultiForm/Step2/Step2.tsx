import { ChangeEvent, FC, useState } from 'react';
import clsx from 'clsx';

import { FormHeader } from '@components/ui/FormHeader';
import { Button } from '@components/ui/Button';
import { Switcher } from '@components/ui/Switcher';
import { Plan } from '@components/ui/Plan';

import { TStep } from '../types';
import { plans } from './constants';
import styles from './Step2.module.scss';

const Step2: FC<TStep> = ({ handleNext, handleBack, data: { step2Data } }) => {
  const [selectedPlan, setSelectedPlan] = useState<string | null>(
    step2Data?.selectedPlan || null
  );
  const [isYearly, setIsYearly] = useState(!!step2Data?.isYearly);

  const handleTogglePeriod = () => {
    setIsYearly(!isYearly);
  };

  const handlePlanChange = (event: ChangeEvent<HTMLInputElement>) => {
    setSelectedPlan(event.target.value);
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    if (!selectedPlan) {
      return;
    }

    handleNext && handleNext({ step2Data: { selectedPlan, isYearly } });
  };

  return (
    <div>
      <FormHeader
        title="Select your plan"
        description="You have the option of monthly or yearly billing."
      />
      <form onSubmit={handleSubmit}>
        <div className={styles.plans}>
          {plans.map((plan) => (
            <Plan
              key={plan.title}
              plan={plan}
              isChecked={selectedPlan === plan.title}
              handlePlanChange={handlePlanChange}
              isYearly={isYearly}
            />
          ))}
        </div>

        <div className={styles.period}>
          <p
            className={clsx(styles.option, {
              [styles.choosen]: !isYearly,
            })}
          >
            Monthly
          </p>
          <Switcher isChecked={isYearly} handleToggle={handleTogglePeriod} />
          <p
            className={clsx(styles.option, {
              [styles.choosen]: isYearly,
            })}
          >
            Yearly
          </p>
        </div>

        <div className={styles.buttons}>
          <Button action="back" onClick={handleBack} />
          <Button type="submit" action="next" />
        </div>
      </form>
    </div>
  );
};

export { Step2 };
