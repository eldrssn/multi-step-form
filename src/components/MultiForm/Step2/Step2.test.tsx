import { fireEvent, render, waitFor } from '@testing-library/react';
import { Step2 } from './Step2';

const renderStep2 = (props = {}) => {
  const defaultProps = {
    handleNext: jest.fn(),
    handleBack: jest.fn(),
    data: { step1Data: null, step2Data: null, step3Data: null },
    ...props,
  };

  return render(<Step2 {...defaultProps} />);
};

describe('Step2', () => {
  test('renders Step2 component correctly', () => {
    const { getByText, container } = renderStep2();

    expect(getByText('Select your plan')).toBeInTheDocument();
    expect(
      getByText('You have the option of monthly or yearly billing.')
    ).toBeInTheDocument();
    expect(getByText('Monthly')).toBeInTheDocument();
    expect(getByText('Yearly')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  const mockHandleNext = jest.fn();

  test('selecting a plan updates state and calls handleNext', async () => {
    const { getAllByTestId, getByText } = renderStep2({
      handleNext: mockHandleNext,
    });

    const plans = getAllByTestId('plan');
    fireEvent.click(plans[0]);
    fireEvent.submit(getByText('Next Step'));

    await waitFor(() => {
      expect(plans[0]).toHaveClass('checked');
      expect(mockHandleNext).toHaveBeenCalledWith(
        expect.objectContaining({
          step2Data: { selectedPlan: 'Arcade', isYearly: false },
        })
      );
    });
  });

  test('checking if special offer exists', () => {
    const { getAllByText } = renderStep2({
      data: {
        step1Data: null,
        step2Data: { selectedPlan: 'Arcade', isYearly: true },
        step3Data: null,
      },
    });

    expect(getAllByText(/2 months free/i)).toHaveLength(3);
  });

  test('checking if special offer no exists', () => {
    const { queryByText } = renderStep2({
      data: {
        step1Data: null,
        step2Data: { selectedPlan: 'Arcade', isYearly: false },
        step3Data: null,
      },
    });

    expect(queryByText(/2 months free/i)).not.toBeInTheDocument();
  });
});
