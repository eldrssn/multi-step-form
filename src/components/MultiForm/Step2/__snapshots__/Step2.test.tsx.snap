// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step2 renders Step2 component correctly 1`] = `
<div>
  <div>
    <h1
      class="title"
    >
      Select your plan
    </h1>
    <p
      class="description"
    >
      You have the option of monthly or yearly billing.
    </p>
    <form>
      <div
        class="plans"
      >
        <label
          class="plan_box"
          data-testid="plan"
        >
          <img
            alt="icon"
            class="icon"
            height="40px"
            src="/images/icon-arcade.svg"
            width="40px"
          />
          <input
            class="box"
            name="plan"
            type="radio"
            value="Arcade"
          />
          <div
            class="description"
          >
            <p
              class="title"
            >
              Arcade
            </p>
            <p
              class="price"
            >
              $
              9
              /
              mo
            </p>
          </div>
        </label>
        <label
          class="plan_box"
          data-testid="plan"
        >
          <img
            alt="icon"
            class="icon"
            height="40px"
            src="/images/icon-advanced.svg"
            width="40px"
          />
          <input
            class="box"
            name="plan"
            type="radio"
            value="Advanced"
          />
          <div
            class="description"
          >
            <p
              class="title"
            >
              Advanced
            </p>
            <p
              class="price"
            >
              $
              12
              /
              mo
            </p>
          </div>
        </label>
        <label
          class="plan_box"
          data-testid="plan"
        >
          <img
            alt="icon"
            class="icon"
            height="40px"
            src="/images/icon-pro.svg"
            width="40px"
          />
          <input
            class="box"
            name="plan"
            type="radio"
            value="Pro"
          />
          <div
            class="description"
          >
            <p
              class="title"
            >
              Pro
            </p>
            <p
              class="price"
            >
              $
              15
              /
              mo
            </p>
          </div>
        </label>
      </div>
      <div
        class="period"
      >
        <p
          class="option choosen"
        >
          Monthly
        </p>
        <label
          class="toggle_switch"
        >
          <input
            class="toggle_input"
            type="checkbox"
          />
          <span
            class="slider"
          />
        </label>
        <p
          class="option"
        >
          Yearly
        </p>
      </div>
      <div
        class="buttons"
      >
        <button
          class="button button_back"
          type="button"
        >
          Go Back
        </button>
        <button
          class="button button_next"
          type="submit"
        >
          Next Step
        </button>
      </div>
    </form>
  </div>
</div>
`;
